Remove support for Python 2 and Python < 3.6

epeisach opened the pull request:
Remove support for Python 2.   This allows additional removal of future and six.   

Update pybind11 to 2.13 and update minimal version of cmake.  For python 3.6/3.7 need pybind11 2.12.1 as latest released pybind has dropped support.

Remove pybind11 2.6.3

Add some extra python versions to tox configuration

Commit:
Drop support for python < 3.6.  Upgrade pybind11, cmake requirements

Commit:
Remove use of future and six

Commit:
Update tox to support more python versions.  Update manifest to include alternative pybind module

Commit:
Bump version to v0.93.0

piehld commented on HISTORY.txt:
```diff
@@ -97,3 +97,5 @@
  7-May-2025    V0.92.0 - BinaryCIF ByteArray Encoder failed to use the provided
                          array type in recorded output type.  Warn deprecation of
 			 obsolete encoders
+12-Jun-2025    V0.93.0 - Remove support for python 2.X, retain support for python 3.6 and newer
```
I feel like this change warrants a major version bump...?

josemduarte replied:
+1

Also with that it'd be nice to start using semantic versioning. Which would be quite useful to users of the library.

piehld commented on setup.py:
```diff
@@ -163,6 +163,7 @@ def build_extension(self, ext):
         "Programming Language :: Python :: 3",
         "Programming Language :: Python :: 3.9",
     ],
+    python_requires=">= 3.6",
```
I think you'll also want to remove the `Python :: 2.7` from the list of classifiers above.

piehld commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
-future
-six
+# future is required by pybind11
+#future
```
Would be good to remove these after confirming things are working

piehld commented on tox.ini:
```diff
@@ -49,10 +49,14 @@ allowlist_externals = echo
+            py38: python3.8
+            py37: python3.7
+            py36: python3.6
             py27: python2.7
```
remove this (py27)?

piehld reviewed (commented):
Thanks @epeisach! I looked over all the files outside of anything under `modules/*`, and think your changes to imports and build files look good. Although, I left a few comments on some minor changes and other things to remove.

We should also think about switching to using a `pyproject.toml` file instead as well, as that is much more modern. See this other user's comment on our `rcsb-api` package: https://github.com/rcsb/py-rcsb-api/issues/67

epeisach commented:
Just seeing the feedback now...
I will make the changes - and test in various scenarios... The future for pybind11 -- I will check the python 3.6/3.7 version to ensure safe.

epeisach commented:
So with regards to pyproject.toml -- we should move in that direction.   I am looking to see if we should remain with setup tools based - or scikit-build-core.    I think that should be another update - so we can get this out.

I will work on the rest today.

piehld commented:
@epeisach If it helps, I just did that recently for `py-rcsb-api`: https://github.com/rcsb/py-rcsb-api/blob/master/pyproject.toml

Commit:
Add pyproject.toml

Update test suite to support python 3.14

Remove mentioned support for python 2.7

Commit:
requirements.txt remove commented out lines

Commit:
Move cibuildwheel.toml into pyproject.toml

Update cibuildwheel to v3.2.1 to build python 3.14 wheels

Use python 3.11 for testing and building wheels as needed by latest cibuildwheel

Commit:
Move all metadata to pyproject.toml leaving setup.py for compilation only

Commit:
Add license back to pyproject.toml

epeisach committed:
Merge pull request #1 from epeisach/pyproject_cleanup

Addition of metadata to pyproject.toml

Commit:
tox.ini: Miss basepython configuration for python 3.14

Commit:
Bump version to 1.0.0

epeisach commented:
The following changes have been done:

a) Version bump to 1.0.0
b) Switch to using pyproject.toml for metadata - stripping down setup.py to the minimal code to build C++ extension
c) Python 2.7 removed from classifiers
d) Testing support (tox) for python 3.13
e) cibuildwheel bumped to new version.  Wheels will be built for python 3.8->14.  This required azure to perform builds with python 3.11 instead of python 3.9
f) For python 3.8/3.9 cibuildwheel will compile wheel with version compatible with glibc 2.27.  This is used by Centos 7 and Ubuntu 18 - which are still in use at RCSB - although could build from sdist
g) The list of files in wheels and source distribution remain the same - except the addition of pyproject.toml in the distributed tar file.

@piehld ready for new review

piehld commented on pyproject.toml:
```diff
@@ -0,0 +1,80 @@
+zip-safe = false
+
+[tool.setuptools.dynamic]
+dependencies = {file = ["requirements.txt"]}
```
Is it possible to remove the `requirements.txt` file and put the dependencies directly in here? Or do you prefer to keep them separate?

Just asking, not necessarily recommending one way or the other. But in my case, I just listed the dependencies in `pyproject.toml` directly since it removed an extra file for me.

epeisach replied:
Yes - we can remove the requirements.txt.  The tox scripts was explicit about including --- but in running the test - it should install as such.   I will test quickly.

piehld replied:
You should be able to replace the references to `requirements.txt` with just a dot in `tox.ini` (or perhaps remove it altogether), like here: https://github.com/rcsb/py-rcsb-api/blob/master/azure-pipelines.yml (in this case, I'm installing it with additional test-suite dependencies, hence the `.[tests]`)

piehld commented on pyproject.toml:
```diff
@@ -0,0 +1,80 @@
+namespaces = false
+
+[tool.setuptools.package-data]
+mmcif = ["*.md", "*.rst", "*.txt", "*.h", "*.C", ".c", "*.cpp"]
```
I don't know about you, but I really struggled to get this part to work as desired. Have you tried building the package locally and checking the contents of the source distribution tarball? If not, I'd suggest doing so and comparing it to what's currently provided on PyPI: https://pypi.org/project/mmcif/#files.

I ended up having to add certain things to the MANIFEST (which it looks like you do) and including the explicit `tool.setuptools.packages.find.include` [setting](https://github.com/rcsb/py-rcsb-api/blob/master/pyproject.toml#L53-L59). I'm still not sure if the `tool.setuptools.package-data` setting I have is actually doing anything or not.

epeisach replied:
Removing these lines makes no differences in the files packaged relative to the purely setup.py packages.  So these lines are now removed.

piehld replied:
I had to do the following to get it to include files in the package: https://github.com/rcsb/py-rcsb-api/blob/master/pyproject.toml#L53-L59

piehld commented on azure-pipelines.yml:
You should probably add a Python 3.14 test too (or are you planning to do that separately?).

epeisach replied:
Will add - but I suspect azure runners do not support yet - but forward looking.

epeisach replied:
I added support to azure-template-tox-job,yaml.

I will see what version is really supported...  Google AI suggests 3.12 but I will test in a separate branch.

piehld replied:
3.14 should be supported—I'm using it here: https://github.com/rcsb/py-rcsb-api/blob/master/azure-pipelines.yml

epeisach replied:
Switched to python 3.14 for building.

piehld commented on pyproject.toml:
```diff
@@ -0,0 +1,80 @@
+archs = ["x86_64", "arm64"]
+
+
+[[tool.cibuildwheel.overrides]]
+select = "*-musllinux*"
+before-all = "apk add cmake bison flex"
+
+
+#[tool.cibuildwheel]
+#manylinux-x86_64-image = "manylinux2014"
+
+[[tool.cibuildwheel.overrides]]
```
Just a general question—why do these headers have double brackets, whereas the others don't?

epeisach replied:
This is the style in the documentation -  search for overrides  https://cibuildwheel.pypa.io/en/v2.19.1/options/

Apparently it is similar to what Mypy did to put Mypy.ini in the pyproject.toml reading some old issues discussion.

piehld commented on tox.ini:
```diff
@@ -49,13 +49,17 @@ allowlist_externals = echo
+            py36: python3.6
 
-[testenv:py27]
+[testenv:py3{6,7,8,9,10,11,12,13,14}]
```
That's cool, I didn't know you could breakup the string like that.

epeisach replied:
Cleaned up.  Remove py27 as a default test.

piehld commented on tox.ini:
```diff
@@ -71,35 +75,14 @@ commands =
-    {envpython} -m unittest discover -v --start-directory {[local_settings]test_path_4} --pattern "{[local_settings]test_pattern}"
+#    {envpython} -m unittest discover -v --start-directory {[local_settings]test_path_2} --pattern "{[local_settings]test_pattern}"
+#    {envpython} -m unittest discover -v --start-directory {[local_settings]test_path_3} --pattern "{[local_settings]test_pattern}"
+#    {envpython} -m unittest discover -v --start-directory {[local_settings]test_path_4} --pattern "{[local_settings]test_pattern}"
```
Maybe take this chance to cleanup these unused lines (both here and above and below).

epeisach replied:
Cleaned up in tox.ini

piehld reviewed (commented):
Thanks @epeisach! Overall I think it looks great. I just had a few comments/suggestions, but nothing major.

Commit:
Remove requirements.txt - move to pyproject,toml.  Remove extraneous optional paths from tox.ini

Commit:
Adjust azure jobs now that requirements.txt no long exists

Commit:
Switch to python 3.14 for testing and building wheels

Commit:
Remove package-data from pyproject.toml as does not affect packaging

piehld commented on setup.py:
So the reason this file can't be removed altogether is because of the lingering cmake setup?

epeisach replied:
That is correct.  We are using the setup tools backend. Metadata has been moved to pyproject.toml - but python code cannot live there. The setup.py is handling all of the cmake invocations, etc.

This would require a change to a backend like scikit-build-core to remove setup.py, leaving the CMakeFile and some more metadata in pyproject,toml.  That is listed in the issue I created yesterday.

piehld replied:
Alright I think I understand better now, thanks.

piehld reviewed (approved):
Thanks @epeisach!
